man/k-shell-doc.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH K-SHELL-DOC-STATIC "1" "May 2020" "koalephant-shell-script-library 2.8.1" "User Commands"
.SH NAME
k-shell-doc-static,  - Generate Markdown documentation from a `shelldoc` documented shell script
.SH SYNOPSIS
.B k-shell-doc-static
[\fI\,OPTION\/\fR]... \fI\,file\/\fR
.SH DESCRIPTION
Generate Markdown documentation from a `shelldoc` documented shell script
.PP
Comment blocks at the beginning of the file, and immediately preceding a function or variable declaration are parsed, reading the following sections: Description, Input, Output, Example, Return and Deprecated.
.PP
The initial lines of the comment block form the description. The other sections must be prefixed with a section delimiter line, e.g. 'Input:' or 'Return:'
.PP
The follow example is from the `k_version_compare` function:
.IP
\f(CW# Compare two version strings\fR
.br
\f(CW#\fR
.br
\f(CW# Version suffixes (e.g. 'alpha', 'rc1' etc) should\fR
.br
\f(CW# use a tilde (~), e.g.: '1.2.0~alpha'\fR
.br
\f(CW# Valid operators are eq/= (equal), neq/!= (not\fR
.br
\f(CW# equal), gt/> (greater than), lt/<  (less than),\fR
.br
\f(CW# gte/>= (greater or equal), lte/<= (less or equal)\fR
.br
\f(CW#\fR
.br
\f(CW# Input:\fR
.br
\f(CW# $1 - the first version to compare\fR
.br
\f(CW# $2 - the second version to compare\fR
.br
\f(CW# $3 - the operator to use, defaults to 'gte'.\fR
.br
\f(CW#\fR
.br
\f(CW# Return:\fR
.br
\f(CW# 0 if the version comparison passes, 1 if not\fR
.PP
The above Shelldoc comment block produces the following Markdown:
.IP
\f(CW### `k_version_compare`\fR
.br
\f(CW\fR
.br
\f(CWCompare two version strings\fR
.br
\f(CWVersion suffixes (e.g. 'alpha', 'rc1' etc) should\fR
.br
\f(CWuse a tilde (~), e.g.: '1.2.0~alpha'\fR
.br
\f(CWValid operators are eq/= (equal), neq/!= (not\fR
.br
\f(CWequal), gt/> (greater than), lt/<  (less than),\fR
.br
\f(CWgte/>= (greater or equal), lte/<= (less or equal)\fR
.br
\f(CW\fR
.br
\f(CW#### Input:\fR
.br
\f(CW * `$1` - the first version to compare\fR
.br
\f(CW * `$2` - the second version to compare\fR
.br
\f(CW * `$3` - the operator to use, defaults to 'gte'.\fR
.br
\f(CW\fR
.br
\f(CW#### Return:\fR
.br
\f(CW0 if the version comparison passes, 1 if not\fR
.SH OPTIONS
.TP
\fB\-D\fR, \fB\-\-debug\fR
Show debug output
.TP
\fB\-h\fR, \fB\-\-help\fR
Show this help
.TP
\fB\-\-md\-extra\-ids\fR bool
Output Markdown Extra compatible syntax to generate ID attributes on Header elements.
Defaults to on in v2.x.x. Will default to off in the next major version.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Suppress all output except errors
.TP
\fB\-V\fR, \fB\-\-verbose\fR
Show verbose output
.TP
\fB\-v\fR, \fB\-\-version\fR
Show the version
.SH COPYRIGHT
Copyright \(co 2020, Koalephant Co., Ltd


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.8.1] - 2020-05-02
### Fixed
- Help output alignment in `k-tool-install`

## [2.8.0] - 2020-04-10
### Added
- `k_tool_description_add`, `k_tool_description_code_add` for adding formatted blocks of text/code to 'help' output.
- `k_tool_usage_text`, `k_tool_usage_code` for formatting blocks of test/code to use in 'help' output.
- `k_tool_environment_add` for adding a defined environment variable to 'help' output
- `k_tool_environment_alias` for adding a defined environment variable alias to 'help' output
- `k_tool_environment_print` for printing defined environment 'help' output 
- `k_function_exists` for checking if a named function is defined

### Fixed
- Allow `k_fs_resolve` to resolve paths to non-existent files/directories 

### Deprecated
- `k_tool_options_text` use `k_tool_description_add` or `k_tool_usage_text` instead
- `k_tool_options_code` use `k_tool_description_code_add` or `k_tool_usage_code` instead

## [2.7.3] - 2020-03-13
### Fixed
- Fix a bug causing `k_fs_resolve` to return some paths ending with `/`
- Fix a hang when running GPG tests with low entropy 

## [2.7.2] - 2019-11-21
### Added
- Helper named 'provision' scripts in the Vagrantfile for `build`, `test` and `test-release`

### Fixed
- Handle lack of TTY device gracefully
- Fix a bug causing `k-script-build` to hang waiting for input when there is no TTY present.
- Resolved a circular dependency issue and an overridden recipe issue in the Makefile when building temporary `k-script-build-static` asset.
- Resolved sed-related issue preventing some string-handling functions working properly on BSD/macOS systems
- Improve fallback temp file handling
- Improved memory-backed tmp file testing on systems without tmpfs filesystems
- Fix GPG tests on macOS

## [2.7.1] - 2019-11-16
### Fixed
- Let `k-tool-install` pass value of `--version-opt` to detected binaries unquoted. 

## [2.7.0] - 2019-11-15
### Added
- `k_int_{min,max,sum,avg}()` functions, to get minimum, maximum, sum and average value from a series, respectively.
- `k_string_indent()` function to indent lines of text.
- `k_string_wrap()` function to word-wrap lines of text.
- `k_string_prefix()` function to apply a prefix to lines of text.
- `k_string_get_lines()` function to retrieve multiple lines from a string
- `k_tool_options_add()` function to add an entry for formatted options help text.
- `k_tool_options_alias()` function to add an alias entry for formatted options help text.
- `k_tool_options_print()` function to print all formatted options help text.
- `k_tool_options_text()` function to print descriptive help text.
- `k_tool_options_code()` function to print example code-like text.
- `k_tool_has_options()` function to determine whether a tool accepts options.
- `k_tty_width()` function to retrieve the column width of the TTY.
 
### Changed
- `k_string_{get_line,pad_right,pad_left,lower,upper,contains,starts_with,ends_with,remove_start,remove_end,trim}()` functions can read from STDIN.
- `k_options_{split,arg_required,arg_optional}()` functions can handle option values that start with a dash by setting `$KOALEPHANT_OPTIONS_ALLOW_DASH`.
- `k_options_arg_required()` function can accept empty-string option values by setting `$KOALEPHANT_OPTIONS_ALLOW_EMPTY`.
- `k_fs_temp_{get_raw,dir,file,predictable_dir,predictable_file}()` functions accept `$KOALEPHANT_TMP_USE_MEM` and `$KOALEPHANT_TMP_REQ_MEM` to try, and require using an in-memory filesystem, respectively.

### Fixed
- Prevent building release file with `unreleased` entry in CHANGELOG.
- Fix docs issue with `k_version_helper` specifying `$0` is input.
- Spelling mistakes in various shelldoc comments.
- A heap of unlikely-but-possible shell quoting issues.
- Ensure quotes and backticks in values are escaped in `k_env_set()`

## [2.6.2] - 2019-07-27
### Fixed
- Ensure dynamically loaded elements of library still work dynamically in a static script
- issue where `k-script-build` would error when processing a script provided via STDIN

## [2.6.1] - 2019-07-26
### Fixed
- missing date for 2.6.0 release
- missing version in generated `install.sh`

## [2.6.0] - 2019-07-26
### Added
- missing Unit tests for `k_gpg_*` and `k_fs_*` functions
- `k_fs_dir_empty()` function to check if a directory is empty
- `k_fs_temp_get_raw()` function to generate temporary names and get the system temp dir using either `mktemp` or `/dev/urandom`
- a generated 'install' script for use in other projects
- `k_hash_*` functions to support hash generation and verification using modules
- `php`, `shasum` and `algosum` modules for `k_hash_*` functions
- `k_config_modules()` function to list loaded config modules
- `k-tool-install` now supports hash verification on source archives
- `k-tool-install` now supports downloading via either `curl` or `wget`, depending on which is available

### Changed
- renamed `php` and `php_ro` config modules to `php-dba` and `php-ini` respectively
- moved config module functions into separate module files
- `k_options_split()`, `k_options_arg_required()` and `k_options_arg_optional()` now use a variable `$OPTION_VALUE_SET` to set/check if an option has been provided a value
- `k_options_arg_required()` now accepts a second boolean parameter, to accept an empty string value. Defaults to false 

### Fixed
- a bug in `k_fs_resolve()` with certain symlink paths
- `k_fs_temp_*()` and `k_fs_predictable_*()` functions now use `k_fs_temp_get_raw()` rather than `mktemp` directly
- a race condition bug in `k_fs_predictable_file()` when generating the predictable file for the first time
- a situation where `k_fs_predictable_file()` returned filenames without creating them
- k-tool-install erroneously always using `--dev` option with `install-deps.sh` scripts

src/k-shell-doc.sh
#!/bin/sh
# Generate Markdown documentation from a 'shelldoc' documented shell script
# Version: PACKAGE_VERSION
# Copyright: 2017, Koalephant Co., Ltd
# Author: Stephen Reay <stephen@koalephant.com>, Koalephant Packaging Team <packages@koalephant.com>
# Todo: Configurable header format. YAML (current), INI, JSON, XML?

# Generate Markdown documentation from a 'shelldoc' documented shell script
#
# Input:
# $1 - file to process
#
# Output:
# Markdown formatted documentation
#
# Example:
# k_shell_doc file1.sh > file1.md
k_shell_doc() {

	set -eu

	. ./base.lib.sh
	. ./bool.lib.sh
	. ./fs.lib.sh
	. ./number.lib.sh
	. ./string.lib.sh

	# shellcheck disable=SC2039
	local KOALEPHANT_TOOL_DESCRIPTION KOALEPHANT_TOOL_VERSION \
		TMP_FUNC_DESCRIPTION TMP_FUNC_INPUT TMP_FUNC_OUTPUT TMP_FUNC_RETURN TMP_FUNC_EXAMPLE TMP_FUNC_DEPRECATED \
		TMP_HEADER_BLOCK TMP_HEADER_DESCRIPTION \
		showVersion=false verbose=false quiet=false debug=false \
		tmpDir headerMode=true funcMode=false prevFuncMode=false mdExtraHeaderIds=true functionSegment='description'


	# shellcheck disable=SC2016
	readonly KOALEPHANT_TOOL_DESCRIPTION='Generate Markdown documentation from a `shelldoc` documented shell script'
	readonly KOALEPHANT_TOOL_VERSION='PACKAGE_VERSION'
	readonly KOALEPHANT_TOOL_ARGUMENTS='file'
	readonly TMP_FUNC_DESCRIPTION='func-description'
	readonly TMP_FUNC_INPUT='func-input'
	readonly TMP_FUNC_OUTPUT='func-output'
	readonly TMP_FUNC_RETURN='func-return'
	readonly TMP_FUNC_EXAMPLE='func-example'
	readonly TMP_FUNC_DEPRECATED='func-deprecated'
	readonly TMP_HEADER_BLOCK='header-block'
	readonly TMP_HEADER_DESCRIPTION='header-desc'
	readonly tmpDir="$(k_fs_temp_dir)"

	k_tool_usage_init () {
		# shellcheck disable=SC2119
		k_tool_description_add <<-'STR'
			Comment blocks at the beginning of the file, and immediately preceding a function or variable declaration are parsed, reading the following sections: Description, Input, Output, Example, Return and Deprecated.

			The initial lines of the comment block form the description. The other sections must be prefixed with a section delimiter line, e.g. 'Input:' or 'Return:'

			The follow example is from the `k_version_compare` function:

		STR

		k_tool_description_code_add 4 '# ' <<-'EXAMPLE'
			Compare two version strings

			Version suffixes (e.g. 'alpha', 'rc1' etc) should use a tilde (~), e.g.: '1.2.0~alpha'
			Valid operators are eq/= (equal), neq/!= (not equal), gt/> (greater than), lt/<  (less than), gte/>= (greater or equal), lte/<= (less or equal)

			Input:
			$1 - the first version to compare
			$2 - the second version to compare
			$3 - the operator to use, defaults to 'gte'.

			Return:
			0 if the version comparison passes, 1 if not
		EXAMPLE

		# shellcheck disable=SC2119
		k_tool_description_add <<-'STR'

			The above Shelldoc comment block produces the following Markdown:

		STR

		k_tool_description_code_add 4 <<-'STR'
			### `k_version_compare`

			Compare two version strings
			Version suffixes (e.g. 'alpha', 'rc1' etc) should use a tilde (~), e.g.: '1.2.0~alpha'
			Valid operators are eq/= (equal), neq/!= (not equal), gt/> (greater than), lt/<  (less than), gte/>= (greater or equal), lte/<= (less or equal)

			#### Input:
			 * `$1` - the first version to compare
			 * `$2` - the second version to compare
			 * `$3` - the operator to use, defaults to 'gte'.

			#### Return:
			0 if the version comparison passes, 1 if not
		STR

		k_tool_options_add 'debug' 'D' '' 'Show debug output'
		k_tool_options_add 'help' 'h' '' 'Show this help'
		k_tool_options_add 'md-extra-ids' '' 'bool' 'Output Markdown Extra compatible syntax to generate ID attributes on Header elements.' \
 													'Defaults to on in v2.x.x. Will default to off in the next major version.'
		k_tool_options_add 'quiet' 'q' '' 'Suppress all output except errors'
		k_tool_options_add 'verbose' 'V' '' 'Show verbose output'
		k_tool_options_add 'version' 'v' '' 'Show the version'
	}


	tmp_string_file() {
		# shellcheck disable=SC2039
		local name

		name="$1"
		printf -- '%s/%s' "${tmpDir}" "${name}"
	}

	tmp_string_get() {
		# shellcheck disable=SC2039
		local name

		name="$1"
		cat "$(tmp_string_file "${name}")"
	}

	tmp_string_append() {
		# shellcheck disable=SC2039
		local name string

		name="$1"
		string="$2"
		printf -- '%s\n' "${string}" >> "$(tmp_string_file "${name}")"
	}

	tmp_string_set() {
		# shellcheck disable=SC2039
		local name string

		name="$1"
		string="$2"
		printf -- '%s\n' "${string}" > "$(tmp_string_file "${name}")"
	}

	tmp_string_clear() {
		# shellcheck disable=SC2039
		local name

		name="$1"
		printf -- '' > "$(tmp_string_file "${name}")"
	}

	tmp_string_exists() {
		# shellcheck disable=SC2039
		local name

		name="$1"
		[ -s "$(tmp_string_file "${name}")" ]
	}


	parse_header_description() {
		tmp_string_append "${TMP_HEADER_DESCRIPTION}" "$(k_string_trim "$1")"
	}

	parse_header_property() {
		tmp_string_append "${TMP_HEADER_BLOCK}" "$(format_header_property "$1")"
	}

	add_header_property() {
		tmp_string_append "${TMP_HEADER_BLOCK}" "$(printf -- '%s: %s' "$@")"
	}

	format_header_property() {
		# shellcheck disable=SC2039
		local line property value

		line="$1"
		property="$(k_string_remove_end ':*' "${line}")"
		value="$(k_string_trim "$(k_string_remove_start "${property}:" "${line}")")"

		printf -- '%s: %s' "$(k_string_lower "$(k_string_trim "${property}")")" "${value}"
	}

	print_function_output() {
		if tmp_string_exists "${TMP_FUNC_OUTPUT}"; then
			printf -- '#### Output:\n%s\n\n' "$(tmp_string_get "${TMP_FUNC_OUTPUT}")"
		fi
	}


	print_headers() {
		# shellcheck disable=SC2039
		local block

		block='---'
		printf -- '%s\n%s\n' "${block}" "$(tmp_string_get "${TMP_HEADER_BLOCK}")"

		if tmp_string_exists "${TMP_HEADER_DESCRIPTION}"; then
			printf -- '%s: %s\n' description  "$(tmp_string_get "${TMP_HEADER_DESCRIPTION}")"
		fi

		printf -- '%s\n\n' "${block}"
	}


	parse_header_line() {
		# shellcheck disable=SC2039
		local line property value

		line="$1"

		if ! k_string_starts_with '#' "${line}"; then
			k_log_info 'End of file header detected'
			print_headers
			headerMode=false
			return
		fi

		line="$(k_string_remove_start '#' "${line}")"

		case "$line" in

			@Ignore*|!*)
			;;

			*:*)
				parse_header_property "${line}"
			;;

			*)
				parse_header_description "${line}"
			;;
		esac
	}


	parse_function_input() {
		tmp_string_append "${TMP_FUNC_INPUT}" "$(format_function_input "$1")"
	}
	parse_function_output() {
		tmp_string_append "${TMP_FUNC_OUTPUT}" "$(format_function_links "$1")"
	}

	parse_function_return() {
		tmp_string_append "${TMP_FUNC_RETURN}" "$(format_function_links "$1")"
	}

	parse_function_description() {
		tmp_string_append "${TMP_FUNC_DESCRIPTION}" "$(format_function_links "$(k_string_trim "${1}")")"
	}

	parse_function_example() {
		tmp_string_append "${TMP_FUNC_EXAMPLE}" "$(format_function_links "$1")"
	}

	parse_function_deprecated() {
		tmp_string_append "${TMP_FUNC_DEPRECATED}" "$(format_function_links "$1")"
	}

	format_function_input() {
		# shellcheck disable=SC2039
		local line name description

		line="$1"
		name="$(k_string_remove_end ' - *' "${line}")"
		description="$(k_string_trim "$(k_string_remove_start "${name} - " "${line}")")"

		# shellcheck disable=SC2016
		printf -- ' * `%s` - %s\n' "$(k_string_trim "$name")" "$(format_function_links "$description")"
	}

	format_function_links() {
		# shellcheck disable=SC2016
		printf -- '%s\n' "$1" | sed -E -e 's/\(([^#]*)#([A-Za-z_]{1,})\)/[`\2`](\1#\2)/g'
	}

	print_function_title() {
		# shellcheck disable=SC2039
		local flags=''

		if [ -n "${3}" ]; then
			#shellcheck disable=SC2016,SC2086
			flags="$(printf -- '`%s` ' "$3")"
		fi

		if [ "${mdExtraHeaderIds}" = true ]; then
			#shellcheck disable=SC2016,SC2086
			printf -- '### `%s` {#%s}\n%s%s\n\n' "$1" "${2:-${1}}" "${flags}" "$(tmp_string_get "${TMP_FUNC_DESCRIPTION}")"
		else
			#shellcheck disable=SC2016,SC2086
			printf -- '### `%s`\n%s%s\n\n' "$1" "${flags}" "$(tmp_string_get "${TMP_FUNC_DESCRIPTION}")"
		fi

	}

	print_function_input() {
		if tmp_string_exists "${TMP_FUNC_INPUT}"; then
			printf -- '#### Input:\n%s\n\n' "$(tmp_string_get "${TMP_FUNC_INPUT}")"
		fi
	}

	print_function_output() {
		if tmp_string_exists "${TMP_FUNC_OUTPUT}"; then
			printf -- '#### Output:\n%s\n\n' "$(tmp_string_get "${TMP_FUNC_OUTPUT}")"
		fi
	}

	print_function_return() {
		if tmp_string_exists "${TMP_FUNC_RETURN}"; then
			printf -- '#### Return:\n%s\n\n' "$(tmp_string_get "${TMP_FUNC_RETURN}")"
		fi
	}

	print_function_deprecated() {
		if tmp_string_exists "${TMP_FUNC_DEPRECATED}"; then
			printf -- '#### Deprecated:\n%s\n\n' "$(tmp_string_get "${TMP_FUNC_DEPRECATED}")"
		fi
	}

	print_function_example() {
		if tmp_string_exists "${TMP_FUNC_EXAMPLE}"; then
			printf -- '#### Example:\n~~~sh\n%s\n~~~\n\n' "$(tmp_string_get "${TMP_FUNC_EXAMPLE}")"
		fi
	}

	parse_body_line() {
		# shellcheck disable=SC2039
		local line paddedLine name identifier='' flags=''

		line="$1"

		if [ "${#line}" -eq 0 ]; then
			return
		fi

		if ! k_string_starts_with '#' "${line}"; then
			if [ "${funcMode}" != true ]; then
				return
			fi

			if k_string_contains '()' "${line}"; then
				name="$(k_string_remove_end '()*' "${line}")"
			else
				identifier="$(k_string_remove_end '=*' "${line}")"
				if k_string_starts_with 'readonly ' "${identifier}"; then
					identifier="$(k_string_remove_start 'readonly ' "${identifier}")"
					flags='readonly'
				elif k_string_starts_with 'local ' "${identifier}"; then
					identifier="$(k_string_remove_start 'local ' "${identifier}")"
				fi

				name="\$${identifier}"
			fi

			print_function_title "${name}" "${identifier}" "${flags}"

			print_function_input
			print_function_output
			print_function_return
			print_function_deprecated
			print_function_example

			prevFuncMode="${funcMode}"
			funcMode=false
			functionSegment=description

			tmp_string_clear "${TMP_FUNC_DESCRIPTION}"
			tmp_string_clear "${TMP_FUNC_INPUT}"
			tmp_string_clear "${TMP_FUNC_OUTPUT}"
			tmp_string_clear "${TMP_FUNC_RETURN}"
			tmp_string_clear "${TMP_FUNC_DEPRECATED}"
			tmp_string_clear "${TMP_FUNC_EXAMPLE}"

		else
			prevFuncMode="${funcMode}"
			funcMode=true
			paddedLine="$(k_string_remove_start '#' "${line}")"
			line="$(k_string_trim "${paddedLine}")"

			if [ "${#line}" -eq 0 ]; then
				return
			fi

			case "${line}" in
				shellcheck*)
					funcMode="${prevFuncMode}"
					return
				;;

				@Ignore*|@TODO*|@Todo*)
					return
				;;

				Input:*)
					tmp_string_clear "${TMP_FUNC_INPUT}"
					functionSegment='input'
					return
				;;

				Output:*)
					tmp_string_clear "${TMP_FUNC_OUTPUT}"
					functionSegment='output'
					return
				;;

				Return:*)
					tmp_string_clear "${TMP_FUNC_RETURN}"
					functionSegment='return'
					return
				;;

				Deprecated:*)
					tmp_string_clear "${TMP_FUNC_DEPRECATED}"
					functionSegment='deprecated'
					return
				;;

				Example:*)
					tmp_string_clear "${TMP_FUNC_EXAMPLE}"
					functionSegment='example'
					return
				;;
			esac
			k_log_info 'Found function segment type: "%s" for "%s"' "${functionSegment}" "${line}"

			case "${functionSegment}" in
				description)
					parse_function_description "${line}"
				;;

				input)
					parse_function_input "${line}"
				;;

				output)
					parse_function_output "${line}"
				;;

				return)
					parse_function_return "${line}"
				;;

				deprecated)
					parse_function_deprecated "${line}"
				;;

				example)
					parse_function_example "$(k_string_remove_start ' ' "${paddedLine}")"
				;;
			esac
		fi
	}

	parse_file() {
		# shellcheck disable=SC2039
		local file line

		file="$1"

		add_header_property 'title' "${file##*/}"

		while read -r line; do
			if [ "${headerMode}" = true ]; then
				parse_header_line "${line}"
			else
				parse_body_line "${line}"
			fi
		done < "${file}"
	}

	k_log_level "${KOALEPHANT_LOG_LEVEL_NOTICE}" > /dev/null
	k_fs_temp_exit

	while k_options_split "$@"; do
		case "$OPTION_NAME" in

			(--md-extra-ids)
				# shellcheck disable=SC2119
				k_options_arg_required
				mdExtraHeaderIds="$(k_bool_parse "${OPTION_VALUE}" true)"
				shift 2
			;;

			(-h|--help)
				k_usage
				exit
			;;

			(-V|--verbose)
				k_log_level "${KOALEPHANT_LOG_LEVEL_INFO}" > /dev/null
				verbose=true
				shift
			;;

			(-D|--debug)
				k_log_level "${KOALEPHANT_LOG_LEVEL_DEBUG}" > /dev/null
				# shellcheck disable=SC2034
				debug=true
				shift
			;;

			(-q|--quiet)
				k_log_level "${KOALEPHANT_LOG_LEVEL_ERR}" > /dev/null
				quiet=true
				shift
			;;

			(-v|--version)
				showVersion=true
				shift
			;;

			(--)
				shift
				break
			;;

			(-*)
				k_log_err 'Unknown option: "%s"' "${OPTION_NAME}"
				k_usage
				return 1
			;;

			(*)
				break
			;;
		esac
	done


	k_version_helper "${showVersion}" "${quiet}" "${verbose}" && return 0

	if [ -z "${1:-}" ]; then
		k_log_err 'No file specified'
		k_usage
		return 1
	fi

	parse_file "$@"
}

k_shell_doc "$@"
